staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise_test.go

```diff
@@ -17,105 +17,148 @@ limitations under the License.
 package promise
 
 import (
-	"sync"
-	"sync/atomic"
+	"context"
 	"testing"
 	"time"
 
-	testclock "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/clock/testing"
+	"k8s.io/apimachinery/pkg/util/wait"
 )
 
-func TestLockingWriteOnce(t *testing.T) {
-	now := time.Now()
-	clock, counter := testclock.NewFakeEventClock(now, 0, nil)
-	var lock sync.Mutex
-	wr := NewWriteOnce(&lock, counter)
-	var gots int32
-	var got atomic.Value
-	counter.Add(1)
+func TestWriteOnceSet(t *testing.T) {
+	oldTime := time.Now()
+	cval := &oldTime
+	ctx, cancel := context.WithCancel(context.Background())
+	wr := NewWriteOnce(nil, ctx.Done(), cval)
+	gots := make(chan interface{})
 	go func() {
-		lock.Lock()
-		defer lock.Unlock()
-
-		got.Store(wr.Get())
-		atomic.AddInt32(&gots, 1)
-		counter.Add(-1)
+		gots <- wr.Get()
 	}()
-	clock.Run(nil)
-	time.Sleep(time.Second)
-	if atomic.LoadInt32(&gots) != 0 {
+	select {
+	case <-gots:
 		t.Error("Get returned before Set")
+	case <-time.After(5 * time.Second):
+		t.Log("Good: Get did not return yet")
 	}
-	func() {
-		lock.Lock()
-		defer lock.Unlock()
-		if wr.IsSet() {
-			t.Error("IsSet before Set")
-		}
-	}()
+	now := time.Now()
 	aval := &now
-	func() {
-		lock.Lock()
-		defer lock.Unlock()
-		if !wr.Set(aval) {
-			t.Error("Set() returned false")
+	if !wr.Set(aval) {
+		t.Error("Set() returned false")
+	}
+	select {
+	case gotVal := <-gots:
+		t.Logf("Got %#+v", gotVal)
+		if gotVal != aval {
+			t.Error("Get did not return what was Set")
 		}
-	}()
-	clock.Run(nil)
-	time.Sleep(time.Second)
-	if atomic.LoadInt32(&gots) != 1 {
+	case <-time.After(wait.ForeverTestTimeout):
 		t.Error("Get did not return after Set")
 	}
-	if got.Load() != aval {
-		t.Error("Get did not return what was Set")
-	}
-	func() {
-		lock.Lock()
-		defer lock.Unlock()
-		if !wr.IsSet() {
-			t.Error("IsSet()==false after Set")
-		}
-	}()
-	counter.Add(1)
 	go func() {
-		lock.Lock()
-		defer lock.Unlock()
-
-		got.Store(wr.Get())
-		atomic.AddInt32(&gots, 1)
-		counter.Add(-1)
+		gots <- wr.Get()
 	}()
-	clock.Run(nil)
-	time.Sleep(time.Second)
-	if atomic.LoadInt32(&gots) != 2 {
-		t.Error("Second Get did not return immediately")
-	}
-	if got.Load() != aval {
-		t.Error("Second Get did not return what was Set")
-	}
-	func() {
-		lock.Lock()
-		defer lock.Unlock()
-		if !wr.IsSet() {
-			t.Error("IsSet()==false after second Get")
+	select {
```

Comment:
nit: let's maybe encapsulate into its own function:

```
func expectGotValue(wr WriteOnce, gotVal <-chan interface{}, expected interface{}) {
  select {
  case gotVal := <-gots:
    if gotVal != aval {
      t.Error("Get returned %v, expected: %v", gotVal, ")
    }
  case <-time.After(wait.ForeverTestTimeout):
    t.Error("Get didn't return")
  }
}
```

And use it in all places [will make tests easier to follow]

Reply:
done